linkerd/app/outbound/src/http/proxy_connection_close.rs

```diff
@@ -46,34 +59,51 @@ where
 {
     type Response = S::Response;
     type Error = S::Error;
-    type Future = Pin<Box<dyn Future<Output = Result<S::Response, S::Error>> + Send + 'static>>;
+    type Future = ResponseFuture<S::Future>;
 
     #[inline]
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         self.inner.poll_ready(cx)
     }
 
+    #[inline]
     fn call(&mut self, req: http::Request<A>) -> Self::Future {
-        let ClientHandle { close, .. } = req
+        let client = req
             .extensions()
             .get::<ClientHandle>()
             .cloned()
             .expect("missing client handle");
+        let inner = self.inner.call(req);
+        ResponseFuture { inner, client }
+    }
+}
 
-        Box::pin(self.inner.call(req).map_ok(move |rsp| {
-            if let Some(msg) = rsp.headers().get(L5D_PROXY_ERROR) {
-                tracing::debug!(
-                    ?msg,
-                    "Received an error response from a peer proxy; closing connection"
-                );
+impl<B, F: TryFuture<Ok = http::Response<B>>> Future for ResponseFuture<F> {
+    type Output = Result<http::Response<B>, F::Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let rsp = futures::ready!(this.inner.try_poll(cx))?;
+
+        if let Some(proxy_conn) = rsp.headers().get(L5D_PROXY_CONNECTION) {
+            if proxy_conn == "close" {
```

Comment:
nit: I _think_ this could just be
```suggestion
        if let Some("close") = rsp.headers().get(L5D_PROXY_CONNECTION) {
```

Reply:
Unfortunately not.
```
  --> linkerd/app/outbound/src/http/proxy_connection_close.rs:88:29
   |
88 |         if Some("close") == rsp.headers().get(L5D_PROXY_CONNECTION) {
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `str`, found struct `http::HeaderValue`
   |
   = note: expected enum `std::option::Option<&str>`
              found enum `std::option::Option<&http::HeaderValue>`
```

`HeaderValue` impls `PartialEq<str>`; but I assume that `Option<T>` doesn't impl `PartialEq<Option<U>> where T: PartialEq<U>`

Reply:
See https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=86bad11540423cd5ece18ed6fb8dbdda